### PR TITLE
Add MSHV support

### PR DESCRIPTION
**Pull request by liuw**
These two patches add device fd support for MSHV to vfio-ioctls.

Note that patch is still needs one modification to point to rust-vmm's repository. It is pointing to a local path for MSHV. I will change it once https://github.com/rust-vmm/mshv/pull/11 is merged.

### CODE REVIEW

**Comment by liuw** [comment]
@andreeaflorescu @lauralt 

The pipeline failed due to mutually exclusive features (`kvm` and `mshv` shouldn't be enabled at the same time). Rust does not support for that use case. What's your thought on augmenting the pipeline for this use case?

I think what I need is:
  - Create a customized pipeline.
  - Disable some tests in the common pipeline.

@jiangliu @sameo @sboeuf 

It would be good if you can have a look at the change before I take on the work to fix the pipeline. If adding mutually exclusive features is not popular I will need to come up with something else.

**Comment by andreeaflorescu** [comment]
@liuw this was a problem that I was facing when implementing mutually exclusive features. The way I fixed it is to choose a default feature when both are specified. Here is the commit that was doing that: https://github.com/rust-vmm/kvm-bindings/commit/190dbaa676db1131731b56935b8cf64f1897df7b

I think it makes sense to not have both features enabled at the same time. At the CI level, you'll also need a custom pipeline to run `cargo test` & `cargo build` with the non-default feature.

File: src/vfio_device.rs
Comment (jiangliu) [inline src/vfio_device.rs:277]: Why is it explicitly named as "mshv_device_add_group()" and "kvm_device_add_group()"?
The function signature seems identical.

File: src/vfio_device.rs
Comment (jiangliu) [inline src/vfio_device.rs:58]: Could we merge "KvmSetDeviceAttr" and "MshvSetDeviceAttr" as "SetDeviceAttr"? It may help to simplify the clients.

File: src/vfio_device.rs
Comment (liuw) [reply src/vfio_device.rs:277]: Using `kvm_device_add_group` for MSHV is not correct and vice versa.

I can rename this function to `device_add_group`, but that will break existing clients because the old function is gone.

If you're okay with the downside, I can make the change.

File: src/vfio_device.rs
Comment (liuw) [reply src/vfio_device.rs:58]: Yes we can. I just didn't want to break existing clients. I can rename `KvmSetDeviceAttr` to `SetDeviceAttr` if people are okay with suddenly losing a variant and hence breaking existing users of this library.

File: src/vfio_device.rs
Comment (andreeaflorescu) [reply src/vfio_device.rs:277]: Shouldn't this function (and the kvm correspondent) need to be marked as `unsafe` because they're taking a `RawFd`? The caller needs to make sure that the fd is corresponding to the proper device.

I would be curious to know what happens when you pass an invalid fd. We could have a negative tests with this scenario (for example passing `-2` as the RawFd.

Comment (liuw) [reply src/vfio_device.rs:277]: Acutally, `kvm_device_add_group` is not public, so removing it should be fine.

@andreeaflorescu it is a private function. Also, the kernel will return an error if an invalid Fd is passed to it. I don't think it needs to be `unsafe`, but I don't have strong opinions here.

Comment (andreeaflorescu) [reply src/vfio_device.rs:277]: In Rust functions that take as a parameter a `RawFd` are typically declared unsafe. I found the explanation here to be the most comprehensive in talking about why it needs to be declared as unsafe: https://github.com/rust-lang/rust/issues/72175#issuecomment-815433842.

Also, since it should be relatively straight forward to test the assumption that kvm returns an error I would suggest to add that as a test case. We've been adding negative tests to kvm-ioctls as well to test various KVM-related assumptions, and this has helped in identifying Linux versions that did not hold those assumptions (https://github.com/rust-vmm/kvm-ioctls/issues/152).

Comment (liuw) [reply src/vfio_device.rs:277]: Sure. I will have a look. Marking it as `unsafe` is easy enough. Let me check what needs to be done for adding some test cases.

Comment (jiangliu) [reply src/vfio_device.rs:277]: The crate hasn't been published yet, so it's acceptable to change APIs:)
@sboeuf what do you think?

Comment (liuw) [reply src/vfio_device.rs:277]: > In Rust functions that take as a parameter a RawFd are typically declared unsafe. I found the explanation here to be the most comprehensive in talking about why it needs to be declared as unsafe: rust-lang/rust#72175 (comment).

There is one simple way of dealing with this safety issue. `device_add_group` doesn't need to take `RawFd` at all. It should be about to take a reference to `VfioGroup` as the second parameter.

File: src/vfio_device.rs
Comment (liuw) [reply src/vfio_device.rs]: I've changed `KvmSetDeviceAttr` to `SetDeviceAttr` in the latest patch.

**Comment by liuw** [comment]
@andreeaflorescu I added a `cunstom-tests.json` to this project. The CI is not picking it up though. I suppose buildkite needs to be configured somehow?

File: .buildkite/custom-tests.json
Comment (andreeaflorescu) [inline .buildkite/custom-tests.json:5]: You should also run clippy on the mshv feature because it is currently running with `--all-features`: https://github.com/rust-vmm/rust-vmm-ci/blob/ae7db2d98a071f52de3d60af9c937204b1f087a4/.buildkite/test_description.json#L41

You'd just need to replace `--all-features` with `--feature mshv`

**Comment by andreeaflorescu** [comment]
> @andreeaflorescu I added a `cunstom-tests.json` to this project. The CI is not picking it up though. I suppose buildkite needs to be configured somehow?

Yes, we need to configure it in Buildkite. Once we do the configuration no other PR will pass the tests until we merge this one. Should we configure it now or do you want to merge anything else before?

**Comment by liuw** [comment]
> > @andreeaflorescu I added a `cunstom-tests.json` to this project. The CI is not picking it up though. I suppose buildkite needs to be configured somehow?
> 
> Yes, we need to configure it in Buildkite. Once we do the configuration no other PR will pass the tests until we merge this one. Should we configure it now or do you want to merge anything else before?

The other in flight PR is your dependabot PR, I suppose it should be rather easy to adapt to the new pipeline configuration?

If you can do it now that would be great, but I'm not sure how you want the other one to be handled.

File: .buildkite/custom-tests.json
Comment (liuw) [reply .buildkite/custom-tests.json:5]: Good suggestion. I've added a new test for that.

**Comment by andreeaflorescu** [comment]
The other PR can wait, we can merge this one first if there are not other PRs in flight. I'll setup buildkite today.

**Review by jiangliu — APPROVED** [review]

File: .buildkite/custom-tests.json
Comment (andreeaflorescu) [inline .buildkite/custom-tests.json:10]: Shouldn't this be:
```cargo clippy --workspace --bins --examples --benches --features=mshv --features mshv --all-targets -- -D warnings``` instead?

**Review by andreeaflorescu — COMMENTED** [review]
I added the Buildkite configuration and now the custom tests are also running: https://buildkite.com/rust-vmm/vfio-ioctls-ci/builds/58#404de176-1fbc-421d-b126-f82a9f18d1b3

File: .buildkite/custom-tests.json
Comment (liuw) [reply .buildkite/custom-tests.json:10]: This crate has defaulted to use `kvm` in `Cargo.toml`. That was introduced to avoid breaking existing users.

So we do need to disable all the default features (`kvm` in this case) and then enable `mshv` explicitly.

The command you wrote runs fine, but because it did not disable `kvm` it would not be testing `mshv` at all.

Comment (andreeaflorescu) [reply .buildkite/custom-tests.json:10]: Right! My bad. LGTM!

**Review by andreeaflorescu — APPROVED** [review]

**Pull request merged**
